Enable docsum to connect with external llm endpoints via helm

devpramod opened the pull request:
## Description

This PR implements support for using external OpenAI-compatible LLM endpoints with DocSum Helm chart while maintaining backward compatibility with existing configurations.

Motivation
Users need the flexibility to connect to external LLM services (like OpenAI or self-hosted OpenAI-compatible endpoints) instead of always relying on the included LLM components. This adds versatility to our Helm charts without disrupting existing functionality.

Key Changes

1. Added `external-llm-values.yaml` configuration file:
- Provides settings for external LLM endpoints (LLM_SERVER_HOST_IP, LLM_MODEL, OPENAI_API_KEY)
- Disables internal LLM services (tgi, vllm, llm-uservice) when using external endpoints

2. Created `templates/external-llm-configmap.yaml`:
- Manages environment variables required for external LLM integration
- Conditionally created only when external LLM is enabled

3. Updated `templates/deployment.yaml`:
- Added conditional logic to use ConfigMap for environment variables when external LLM is enabled
- Maintained all existing functionality and conditional paths for internal LLM services

3. Updated `Chart.yaml`:
- Added proper conditions for all dependencies to make them optional
- Enables selective component deployment based on configuration values

4. Updated `README.md`:
- Added example commands showing how to use external LLM endpoints
- Maintained consistent placeholder values across all services

## Issues


## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

'n/a'

## Tests

Tested backward compatibility to ensure existing helm charts with default values.yaml is working

eero-t commented:
I guess external test needs the e2e PR to be merged first.

ROCM tests fail to `helm: command not found` in `helm install` test phase  => @chensuyue ?

DocSum Gaudi values failure is a know bug, for which @lianhao used #1132 as a workaround (with another chart): https://github.com/opea-project/GenAIComps/issues/1719

eero-t commented on helm-charts/docsum/README.md:
```diff
@@ -26,7 +26,9 @@ helm install docsum docsum --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} --se
 # To use AMD ROCm device with TGI
 # helm install docsum docsum --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} --values docsum/rocm-tgi-values.yaml
-
+# helm install docsum docsum --set global.HUGGINGFACEHUB_API_TOKEN=${HFTOKEN} --values docsum/gaudi-tgi-values.yaml ..
```
Extra line duplicating one from above?

eero-t replied:
There's now another duplicate line, but as Lianhao commented also, I'll resolve this.

lianhao commented:
Please rebase after https://github.com/opea-project/GenAIInfra/pull/1146 is merged, for the default model changes in DocSum

srinarayan-srikanthan commented on helm-charts/docsum/README.md:
```diff
@@ -26,6 +26,10 @@ helm install docsum docsum --set global.HF_TOKEN=${HFTOKEN} --set global.modelUs
 # To use AMD ROCm device with TGI
 # helm install docsum docsum --set global.HF_TOKEN=${HFTOKEN} --values docsum/rocm-tgi-values.yaml
+# To use Gaudi device with TGI
+# helm install docsum docsum --set global.HF_TOKEN=${HFTOKEN} --values docsum/gaudi-tgi-values.yaml ..
```
done

lianhao reviewed (approved):
code LGTM. Let's wait the CI back online before merging this.

eero-t reviewed (commented):
(Not setting this review as change request as I will be OoO and cannot approve fixes to them before code freeze.)

eero-t commented:
> I would suggest using LLM_SERVICE_HOST as the Helm variable name even if it would mapped to LLM_SERVICE_HOST_IP environment variable ...
> 
> => I'll add PR for doing that change for already merged ChatQnA & CodeGen Helm charts (+ add KubeAI example).

Half of _GenAIExamples_ apps use `LLM_SERVER_` and half use `LLM_SERVICE_` prefix for env variable names (see https://github.com/opea-project/GenAIExamples/issues/2143).

While most of apps in _GenAIInfra have Helm charts belong to latter category, I decided to use the shorter prefix for Helm variable names (`LLM_SERVER_HOST` / `LLM_SERVER_PORT`).

=> Please see PR #1166

eero-t reviewed (changes requested):
> I am running it something like: helm install docsum docsum --set global.HF_TOKEN=${HFTOKEN} --set global.modelUseHostPath=${MODELDIR} --set llm-uservice.LLM_MODEL_ID=${MODELNAME} --set vllm.LLM_MODEL_ID=${MODELNAME}

You're running the whole OPEA DocSum application at the external / remote end, to provide remote `llm-uservice`?

> Then port-forwarding kubectl port-forward svc/docsum 8888:8888 to test the application.

Err... I assume you mean to the local DocSum instance, not to the external/remote one?

----

I discussed this with Sakari, and supporting enterprise inferencing services (i.e. OpenAI APIs instead of OPEA wrapper services) can be done in separate PR.

However, needing external end point to be the wrapper service instead of LLM itself, means that this PR needs several changes:
* Example host names: "llm-server" -> "llm-uservice"
* Comments: "OpenAI LLM" -> "OPEA DocSum LLM wrapper (llm-uservice)"
* Drop KubeAI example as it provides just standard OpenAI LLM API, not DocSum LLM wrapper one
* Drop `OPENAI_API_KEY` variable as it's not relevant for this

=> it may be easier to change DocSum `llm-uservice` service to use external OpenAI LLM, but you need to test that it works.

louie-tsai commented:
@louie-tsai  need to review it.

devpramod commented:
@eero-t You're absolutely right, and I apologize for the oversight. I initially thought DocSum could work like ChatQnA (which calls /v1/chat/completions directly), but I now understand that DocSum specifically requires the
  /v1/docsum endpoint that only llm-uservice provides.

  The original implementation was misleading - it suggested users could connect directly to external LLMs when they really needed another llm-uservice instance.

  I've updated the PR to follow your suggested approach:
  - Keep llm-uservice enabled (as the required wrapper for /v1/docsum)
  - Configure llm-uservice itself to use external LLM endpoints (OpenAI, vLLM, TGI, etc.)
  - Removed the confusing environment variables from DocSum that it doesn't actually use

  This provides the external LLM functionality users want while maintaining the proper OPEA architecture. Thank you for the detailed feedback - it helped clarify the distinction between DocSum's requirements and ChatQnA's more
  flexible endpoint usage.

eero-t commented on helm-charts/docsum/variant_external-llm-values.yaml:
```diff
@@ -0,0 +1,19 @@
+    LLM_ENDPOINT: "https://api.openai.com/v1"  # External LLM API endpoint (OpenAI, vLLM, TGI, etc.)
+    OPENAI_API_KEY: "${OPENAI_API_KEY}"        # API key for authentication
+    LLM_MODEL_ID: "gpt-4-turbo"               # Model to use
+    TEXTGEN_BACKEND: "openai"                  # Backend type for OpenAI-compatible endpoints
```
This is DocSum service, not TextGen.

Also, it looks that giving anything else than `TGI` or `vLLM` to the `*_BACKEND` variables is going to fail Helm install: https://github.com/opea-project/GenAIInfra/blob/main/helm-charts/common/llm-uservice/templates/configmap.yaml

(In `TEXTGEN_BACKEND` case there seems to be also `BEDROCK` option, but that does not help.)

Looking at `DOCSUM_BACKEND="vLLM"` backend implementation, it seems like it would be OpenAI API compatible, but it hard-codes `openai_api_key`  setting to `EMPTY` string: https://github.com/opea-project/GenAIComps/blob/main/comps/llms/src/doc-summarization/integrations/vllm.py#L58 :-/

devpramod replied:
@eero-t Based on Sri's comment I have fallen back to docusm backend with vLLM, please review.

chensuyue commented:
@devpramod does this PR still target v1.4?

devpramod commented on helm-charts/docsum/Chart.yaml:
```diff
@@ -21,13 +21,16 @@ dependencies:
   - name: llm-uservice
     version: 0-latest
     repository: "file://../common/llm-uservice"
+    condition: llm-uservice.enabled
   - name: whisper
     version: 0-latest
     repository: "file://../common/whisper"
+    condition: whisper.enabled
   - name: ui
     version: 0-latest
     repository: "file://../common/ui"
     alias: docsum-ui
+    condition: docsum-ui.enabled
```
Hi @eero-t The conditional enablement is necessary for configurability. For example,  in our productivity suite terraform project, we need to deploy only the core DocSum service without the UI or other optional components.

eero-t commented on helm-charts/docsum/README.md:
```diff
@@ -26,7 +26,12 @@ helm install docsum docsum --set global.HF_TOKEN=${HFTOKEN} --set global.modelUs
 # To use AMD ROCm device with TGI
 # helm install docsum docsum --set global.HF_TOKEN=${HFTOKEN} --values docsum/rocm-tgi-values.yaml
-
+# To use with external OpenAI-compatible LLM endpoints (OpenAI, vLLM, TGI, etc.)
+# This configures the llm-uservice to connect to external LLM providers while maintaining DocSum compatibility
+# For OpenAI:
+# helm install docsum docsum --values docsum/variant_external-llm-values.yaml --set llm-uservice.OPENAI_API_KEY="your-api-key" --set llm-uservice.LLM_ENDPOINT="https://api.openai.com" --set llm-uservice.LLM_MODEL_ID="gpt-4-turbo"
+# For vLLM/TGI or other OpenAI-compatible endpoints:
+# helm install docsum docsum --values docsum/variant_external-llm-values.yaml --set llm-uservice.LLM_ENDPOINT="http://your-server-url" --set llm-uservice.LLM_MODEL_ID="your-model"
```
Could you do separate PR adding similar stuff for FaqGen in ChatQnA Helm chart?

(Separate variant file, README instructions etc.)

eero-t commented on helm-charts/docsum/variant_external-llm-values.yaml:
```diff
@@ -0,0 +1,17 @@
+# Copyright (C) 2024 Intel Corporation
```
```suggestion
# Copyright (C) 2025 Intel Corporation
```

eero-t reviewed (approved):
Approved, seems OK now.

chensuyue commented:
Please note that the CI/CD machines will be taken back on 8/25, and it is not yet clear when the new machines will be assigned.
